01-data-structures/06-trees/min_binary_heap/min_binary_heap.rb

```diff
@@ -0,0 +1,134 @@
+require_relative "node"
+
+class MinBinaryHeap
+  attr_reader :root
+  attr_reader :tree
+
+  def initialize(root)
+    @root = root.clone
+    @tree = [nil, @root]
+  end
+
+  def insert(root, data)
+    @tree.push(data.clone)
+    index = @tree.length - 1
+    parent = index / 2
+
+    if @tree[parent] && index % 2 == 0 
+      @tree[parent].left = @tree[index]
+    elsif @tree[parent] && index % 2 != 0 
+      @tree[parent].right = @tree[index]
+    end
+    parent_is_greater_than_child = @tree[parent] && @tree[parent].rating > @tree[index].rating
+    while parent_is_greater_than_child
+      temp = @tree[parent].clone
+      @tree[parent].title = @tree[index].title
+      @tree[parent].rating = @tree[index].rating
+      @tree[index].title = temp.title
+      @tree[index].rating = temp.rating
+      index = parent
+      parent = index / 2
+      parent_is_greater_than_child = @tree[parent] && @tree[parent].rating > @tree[index].rating
+    end
+  end
+
+  def find(root, data)
+    (1...@tree.length).each do |location|
+      if data.title == @tree[location].title
+        return @tree[location]
+      end    
+    end
+    nil
+  end
+
+  def delete(root, data)
```

Comment:
Basically this is your pseudocode:
 1, Delete a node from the array 
       (this creates a "hole" and the tree is no longer "complete")

    2. Replace the deletion node
       with the "fartest right node" on the lowest level
       of the Binary Tree
       (This step makes the tree into a "complete binary tree")

    3. Heapify (fix the heap):

         if ( value in replacement node < its parent node )
            Filter the replacement node UP the binary tree
         else
	    Filter the replacement node DOWN the binary tree